spark3/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanUtils.scala

```diff
@@ -54,14 +55,28 @@ object PlanUtils {
         Integer.TYPE)
       .build()
 
+  def numShufflePartitions(conf: SQLConf): Option[Int] = {
+    if (Spark3VersionUtil.isSpark30) {
+      Some(conf.numShufflePartitions)
+    } else {
+      // SPARK-32056: Coalesce partitions for repartition by expressions when AQE is enabled
+      None
+    }
+  }
+
   def createRepartitionByExpression(
       partitionExpressions: Seq[Expression],
       child: LogicalPlan,
-      numPartitions: Int): RepartitionByExpression = {
+      numPartitions: Option[Int]): RepartitionByExpression = {
     if (Spark3VersionUtil.isSpark30) {
-      repartitionByExpressionCtor.newInstance(partitionExpressions, child, Integer.valueOf(numPartitions))
+      numPartitions match {
+        case Some(num) =>
+          repartitionByExpressionCtor.newInstance(partitionExpressions, child, Integer.valueOf(num))
+        case None =>
+          throw new IllegalArgumentException("numPartitions is required before SPARK-32056")
```

Comment:
It doesn't make sense to me that there is a case where we would allow passing an incorrect value here. Rather than creating the `numShufflePartitions` method, checking for Spark 3.0 in both places, and then adding this check, why not just change the behavior so that the number of shuffle partitions is an int, but is ignored unless this is Spark 3.0?

Reply:
> It doesn't make sense to me that there is a case where we would allow passing an incorrect value here.

The check is redundant for now, because we only use it in one place, and make sure to call it correctly. But I think `createRepartitionByExpression` is likely to be used in the future, add a check here makes it safer.

> any version other that 3.0 will not set shuffle partitions

It's not correct after SPARK-33779 (part of Spark 3.2), https://github.com/apache/iceberg/pull/2512#discussion_r658040759

Are we intend to reuse the code for Spark 3.2?

Reply:
The logic I'm suggesting is a simpler way to get the same behavior here. I think we should go with that for now and update for 3.2 later since we will need to anyway. That isn't changed by my suggestion.

Reply:
Updated as your suggestion.